llvm/lib/Transforms/Vectorize/VPlan.h

```diff
@@ -2239,22 +2239,21 @@ class VPInterleaveRecipe : public VPRecipeBase {
 /// a vector operand into a scalar value, and adding the result to a chain.
 /// The Operands are {ChainOp, VecOp, [Condition]}.
 class VPReductionRecipe : public VPRecipeWithIRFlags {
-  /// The recurrence decriptor for the reduction in question.
-  const RecurrenceDescriptor &RdxDesc;
+  /// The recurrence kind for the reduction in question.
+  RecurKind RdxKind;
   bool IsOrdered;
   /// Whether the reduction is conditional.
   bool IsConditional = false;
 
 protected:
-  VPReductionRecipe(const unsigned char SC, const RecurrenceDescriptor &R,
-                    Instruction *I, ArrayRef<VPValue *> Operands,
-                    VPValue *CondOp, bool IsOrdered, DebugLoc DL)
-      : VPRecipeWithIRFlags(SC, Operands,
-                            isa_and_nonnull<FPMathOperator>(I)
-                                ? R.getFastMathFlags()
-                                : FastMathFlags(),
-                            DL),
-        RdxDesc(R), IsOrdered(IsOrdered) {
+  VPReductionRecipe(const unsigned char SC, RecurKind RdxKind,
+                    FastMathFlags FMFs, Instruction *I,
+                    ArrayRef<VPValue *> Operands, VPValue *CondOp,
+                    bool IsOrdered, DebugLoc DL)
+      : VPRecipeWithIRFlags(
+            SC, Operands,
+            isa_and_nonnull<FPMathOperator>(I) ? FMFs : FastMathFlags(), DL),
```

Comment:
This looks a little odd now. Can we just pass in `FMFs` directly now? I'd expect that if there are no flags then `FMFs` should also be empty. Also, I can imagine a user might want to set the flags without passing an instruction.

Reply:
For now, FMFs are only supported for FPMath ops, and the RecurrenceDescriptor's fast-math flags are set to fast by default for non-FP ops. Arguably we may want to fix this separately. When updating to not pass the recurrence descriptor but FMFs, we should also move the check to the caller.

Reply:
This isa_and_nonnull<FPMathOperator> check was just added in #130881 so that non-fp RecurKinds don't end up setting the flags, because by default a non-fp RecurrenceDescriptor will have all FastMathFlags set, see https://github.com/llvm/llvm-project/pull/130881#discussion_r1998522277

It is a bit weird though I'll give you that. I can try and move this to the call site, would that be cleaner?

Reply:
I've moved it to the caller in c2440656af04ddc5c379979d55b34003f161158e